MudPicker: Add parameter ImmediateText to update Text while typing (#7390)

dennml opened the pull request:
Add parameter ImmediateText to update Text while typing.

This means that StringValueChanged and TextChanged are invoked as the user types a value. 

Without this change, StringValueChanged and TextChanged are only invoked if a Mask has been defined or when the user changes focus to leave the component.

<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
Fixes #7390
Helps with a workaround for #6796 where the caret may jump around when using MudMask

Added a new section **Text Parsing** to DatePicker in Docs. 
![fix2](https://github.com/MudBlazor/MudBlazor/assets/20588465/d28ca739-a2d3-450c-94b9-e0a3c4af4fcd)

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
Tested using Docs for MudDatePicker both with and without a mask defined.

As per #7390 the snippet at https://try.mudblazor.com/snippet/QkQRkimvQnViHRBz has been incorporated into new **Text Parsing** section in Docs.

In the supplied try.mudblazor, type 26.8.23 into the 4th field with label No Mask.
The value will only be set after focus leaves the field in the current version.
(click to view)
![issue](https://github.com/MudBlazor/MudBlazor/assets/20588465/8a4432e1-92da-4a9b-8a68-bb633dcc34f9)


After the change, working with the new example in Docs, the value will be set immediately when the user types a valid date. 
This occurs manually with validation/parsing in TextChanged.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

Commit:
MudPicker: Use Editable to set MudTextField Immediate and invoke TextChanged

Commit:
MudPicker: Add parameter ImmediateText

Commit:
Revert whitespace changes

Commit:
Typo

dennml commented:
@ScarletKuro I see you're active on recent pull requests so I'm guessing you're the best contact. 
Do you have any thoughts on this PR?
Thanks

henon commented:
@dennml can you please record a gif with ScreenToGif and post so we can see how it looks now?

dennml commented:
@henon I've added before and after recordings as requested

ScarletKuro commented:
I think it would be necessary to add here one bUnit test for this.

ScarletKuro commented on src/MudBlazor.Docs/Pages/Components/DatePicker/Examples/DatePickerMaskExample.razor:
```diff
@@ -1,11 +1,54 @@
+        return value.HasValue ? "set" : "not set";
+    }
+
+
```
nit: remove the extra new line

dennml replied:
Reverted changes to this example

henon commented:
We are trying not to overload our docs examples. In this case the "No mask" picker feels like the fifth wheel on the wagon in this example and raises more questions than it answers in this context. I'd opt for either removing the changes to that example altogether or moving this use-case into its own section with a description that explains the new parameter and what you can do with it. Then the example will be much more helpful.

Also, we need test covereage of the new parameter. This means a bUnit test that fails without the parameter and passes with it. This will make sure the feature won't break in the future when others touch the DatePicker component.

Commit:
Revert Docs change

Commit:
Test: Ensure MudPicker has the ImmediateText parameter

Commit:
Add Text Parsing example to Docs

dennml commented:
@henon thanks for the review.

> the "No mask" picker feels like the fifth wheel

I agree. Have reverted the change to Docs to remove "No Mask" from the mask example. 

Added a section **Text Parsing** to Docs for DatePicker and updated the description in the PR. Having this in Docs could be useful for people having difficulties with masking.

> a bUnit test that fails without the parameter and passes with it.

Test added to ensure the ImmediateText parameter exists. I was unsure what tests to add since I couldn't see how to reproduce the actions of user typing. Hopefully the test I added is what you are asking for. It will cause an error if someone were to delete the ImmediateText parameter.

Please let me know if it needs anything else.

henon commented:
Here is one way you can simulate user input: `comp.Find("input").KeyDown(new KeyboardEventArgs() { Key = "a", Type = "keydown", });`

Of course it would be best to make sure the functionality of the parameter is protected as well.

Commit:
Add test: TextChanged with ImmediateText

dennml commented:
Thanks @henon that was helpful. I added another test that ensures the TextChanged callback works when ImmediateText and Editable are both true but not when either is false.

henon commented:
Thx @dennml !
